### PR TITLE
Implement `signal_state` for block encodings

### PR DESCRIPTION
**Pull request by charlesyuan314**
For `LinearCombination`, `Product`, and `ChebyshevPolynomial`, the signal state is `PrepareIdentity` and the bloq requires all constituent block encodings to also have `PrepareIdentity`. This is because the circuits implementing these block encodings either control or reflect on the ancilla register in a way that does not work if the signal state is not zero.

For `Unitary`, `LCU`, and `SparseMatrix`, the signal state is `PrepareIdentity` and there are no constituent block encodings.

For `Phase` and `TensorProduct`, the signal state is derived from the constituent block encoding. The case for `TensorProduct` where the constituent does not have `PrepareIdentity` will be implemented in a follow-up PR.

TODO:
- [x] add tests for signal state
- [x] add test for checks in `ChebyshevPolynomial`

### CODE REVIEW

File: qualtran/bloqs/block_encoding/chebyshev_polynomial.py [hunk 1]
```diff
@@ -75,6 +76,10 @@ class ChebyshevPolynomial(BlockEncoding):
     def __attrs_post_init__(self):
         if self.order < 0:
             raise ValueError(f"order must be greater >= 0. Found {self.order}.")
+        if not isinstance(self.block_encoding.signal_state.prepare, PrepareIdentity):
+            raise ValueError(
+                "Cannot take Chebyshev polynomial of block encodings with non-zero signal state."
+            )
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/chebyshev_polynomial.py:82]: Is this also the case for `ScaledChebyshevPolynomial`?

File: qualtran/bloqs/block_encoding/product_test.py [hunk 2]
```diff
@@ -63,11 +68,41 @@ def test_product_signature():
...
+
+    @property
+    def signature(self) -> Signature:
+        return Signature.build(system=1, ancilla=1, resource=1)
+
+    @property
+    def signal_state(self) -> BlackBoxPrepare:
+        return BlackBoxPrepare(TestPrepareOracle())
+
+
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/product_test.py:98]: move them to `bloqs/for_testing/`?

File: qualtran/bloqs/block_encoding/unitary.py [hunk 3]
```diff
@@ -74,10 +75,7 @@ def pretty_name(self) -> str:
 
     @property
     def signal_state(self) -> BlackBoxPrepare:
-        # This method will be implemented in the future after PrepareOracle
-        # is updated for the BlockEncoding interface.
-        # Github issue: https://github.com/quantumlib/Qualtran/issues/1104
-        raise NotImplementedError
+        return BlackBoxPrepare(PrepareIdentity((QAny(self.ancilla_bitsize),)))
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/unitary.py:78]: Does this cause any issues when `ancilla_bitsize=0`?

**Review by anurudhp — APPROVED** [review]
lgtm! % nits

File: qualtran/bloqs/block_encoding/unitary.py (hunk 3 above)
Comment (charlesyuan314) [reply qualtran/bloqs/block_encoding/unitary.py:78]: It would fail if we ever took that `signal_state` and did a decomposition. It's not something we do currently. 

I think a simple fix would be to add a check in `PrepareIdentity.build_composite_bloq` to not do anything if the bitsize is zero.

File: qualtran/bloqs/block_encoding/chebyshev_polynomial.py (hunk 1 above)
Comment (charlesyuan314) [reply qualtran/bloqs/block_encoding/chebyshev_polynomial.py:82]: It is captured when `ScaledChebyshevPolynomial` instantiates `ChebyshevPolynomial` with the same arguments.

I added the redundant check here though for clarity.

File: qualtran/bloqs/block_encoding/product_test.py (hunk 2 above)
Comment (charlesyuan314) [reply qualtran/bloqs/block_encoding/product_test.py:98]: I could try in a follow-up PR. The thing is that these really are stub classes whose only purpose is to test edge cases of constructors. They're not really useful for independent testing of any other behavior.

File: qualtran/bloqs/block_encoding/unitary.py (hunk 3 above)
Comment (charlesyuan314) [reply qualtran/bloqs/block_encoding/unitary.py:78]: If that (or some other change to `PrepareIdentity`) is desirable we can do in a follow-up PR.
